src/behavior-considered-undefined.md

```diff
@@ -23,44 +23,56 @@ code.
 </div>
 
 * Data races.
-* Dereferencing a null or dangling raw pointer.
-* Unaligned pointer reading and writing outside of [`read_unaligned`]
-  and [`write_unaligned`].
-* Reads of [undef] \(uninitialized) memory.
-* Breaking the [pointer aliasing rules] on accesses through raw pointers;
-  a subset of the rules used by C.
-* `&mut T` and `&T` follow LLVM’s scoped [noalias] model, except if the `&T`
-  contains an [`UnsafeCell<U>`].
-* Mutating non-mutable data &mdash; that is, data reached through a shared
+* Dereferencing (using the `*` operator on) a dangling or unaligned raw pointer.
+* Breaking the [pointer aliasing rules]. `&mut T` and `&T` follow LLVM’s scoped
+  [noalias] model, except if the `&T` contains an [`UnsafeCell<U>`].
+* Mutating non-mutable data (that is, data reached through a shared
   reference or data owned by a `let` binding), unless that data is contained
   within an [`UnsafeCell<U>`].
-* Invoking undefined behavior via compiler intrinsics:
-  * Indexing outside of the bounds of an object with [`offset`] with
-    the exception of one byte past the end of the object.
-  * Using [`std::ptr::copy_nonoverlapping_memory`], a.k.a. the `memcpy32`and
-    `memcpy64` intrinsics, on overlapping buffers.
-* Invalid values in primitive types, even in private fields and locals:
-  * Dangling or null references and boxes.
+* Invoking undefined behavior via compiler intrinsics.
+* Executing code compiled with platform features that the current platform
+  does not support (see [`target_feature`]).
+* Unwinding into another language.
```

Comment:
Wouldn't this be UB in that other language, depending on that languages rules ?

Reply:
Would it? I am not sure. Are you expecting other languages to have a clause like "it is UB to unwind *from* another language"? Rust does not have such a clause.

Reply:
Unwinding from another language would also be ok, as long as the unwind is identical to what Rust uses. E.g. in

```rust
extern "Rust" {
    fn foo();
}
unsafe { foo() }
```

`foo` can unwind there without issues, and whether `foo` is actually implemented in Rust, or say assembly, doesn't matter as long as it follows the Rust ABI.

Reply:
> Would it? I am not sure. Are you expecting other languages to have a clause like "it is UB to unwind from another language"? Rust does not have such a clause.

Rust says that unwinding from Rust into Rust is ok. The behavior of unwinding from some other language into Rust is not defined. Most other languages do this.

Reply:
Maybe you are confusing this with unwinding from `nounwind` functions is UB ?

For example, we do explicitly say that if an `extern "C"` function declaration unwinds, the behavior is undefined. We assume that this does not happen, and e.g. we give these functions the `nounwind` attribute in LLVM-IR.

Reply:
I am just copying this from the Nomicon, where this line existed [for quite some time already](https://doc.rust-lang.org/1.30.0/nomicon/what-unsafe-does.html).

I think it is crucial that the case of unwinding out of an `extern "C"` is covered here somehow. Any suggestions for better wording?

Reply:
Unwinding from extern function declarations whose ABI does not support unwinding, like `extern "C"`, is UB.

Reply:
Emphasis on "some ABIs" and "function declarations". Trying to unwind from an `extern "C"` function definition panics, and trying to unwind from an `extern "Rust"` declaration is ok.

Reply:
> Trying to unwind from an extern "C" function definition aborts

AFAIK it doesn't, not yet? See https://github.com/rust-lang/rust/pull/62505.
And even then there will be an attribute to opt ouf of this, with a chance for people to use it the wrong way.

I was asking for a wording for the "unwinding out of Rust to another language" case. I think that is the more interesting one.

Reply:
Also this code looks a lot like unwinding is UB even when it comes from an `extern "Rust"` *declaration*:

https://github.com/rust-lang/rust/blob/d760df5aea483aae041c9a241e7acacf48f75035/src/librustc_codegen_llvm/attributes.rs#L278-L281

Reply:
> AFAIK it doesn't, not yet? See rust-lang/rust#62505.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8665db1c91123fdaa3617ca8443d6989

That code aborts when the panic tries to escape `foo` :/

Reply:
We abort-on-panic-from-extern only on nightly, but regularly backport a patch reverting that to beta. It is quite silly. :/

See https://github.com/rust-lang/rust/pull/62603.

Reply:
> And even then there will be an attribute to opt of of this, 

AFAIK what will happen here is still very unclear. Chances are there will just be different function ABIs w/o unwinding support, to make that part of the type system.

> I was asking for a wording for the "unwinding out of Rust to another language" case.

Whether you can or cannot unwinding out of an `extern "ABI" fn` depends on the function ABI - some allow this, some do not. The ones that do not, panic. 

When you call a function (from any language) you must uphold its ABI - if you don't, the behavior is undefined. 

For example, if you have a Rust function that unwinds, and call it from Rust code compiled with a slightly different toolchain that happens to be ABI incompatible, the behavior is undefined. Same language, but unwinding triggered UB. If you have a Rust function that unwinds, and after unwinding we land in a different programming language than Rust, saying that this is UB is possible, but kind of pointless, because the program execution isn't in Rust code anymore. It's in some other programming language that might say that unwinding into it from Rust is ok. Either way, we can't tell, and even if we could, I don't know which purpose does this serve.

Reply:
> We abort-on-panic-from-extern only on nightly, but regularly backport a patch reverting that to beta. It is quite silly. :/

I haven't used anything but nightly for so long that I forgot that this is still a thing. So yeah, this should mention this.

Reply:
We agreed on chat that for now, the reference should say that cross-lang unwinding is just plain UB (following the Nomicon), and should be updated once https://github.com/rust-lang/rfcs/pull/2699 is accepted.